apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx

```diff
@@ -0,0 +1,251 @@
+  const [itemToClone, setItemToClone] = useState(null);
+
+  const [itemToDelete, setItemToDelete] = useState(null);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [numPages, setNumPages] = useState(1);
+  const [error, setError] = useState(null);
+
+  const actionsCellFormatter = (actions, {rowData}) => {
+    return (
+      <div style={styles.actionsColumn}>
+        <Button
+          icon="edit"
+          text=""
+          href={rowData.editPath}
+          target="_blank"
+          __useDeprecatedTag
+          color="teal"
+        />
+        <Button
+          onClick={() => setItemToClone(rowData)}
+          text=""
+          icon="clone"
+          color="gray"
+          style={{margin: 0}}
+        />
+        <Button
+          onClick={() => setItemToDelete(rowData)}
+          text=""
+          icon="trash"
+          color="red"
+          style={{margin: 0}}
+        />
+      </div>
+    );
+  };
+
+  const fetchExpressions = (environmentId, categoryId, page) => {
+    const data = {};
+    if (environmentId !== DEFAULT_VALUE) {
+      data.programmingEnvironmentId = environmentId;
+    }
+    if (categoryId !== DEFAULT_VALUE) {
+      data.categoryId = categoryId;
+    }
+    data.page = page;
+    const url =
+      '/programming_expressions/get_filtered_expressions?' +
+      queryString.stringify(data);
+    let success = false;
+    fetch(url)
+      .then(response => {
+        if (response.ok) {
+          success = true;
+        } else {
+          setError(response.statusText);
+        }
+        return response.json();
+      })
+      .then(data => {
+        if (success) {
+          setProgrammingExpressions(data.expressions);
+          setNumPages(data.numPages);
+        }
+      });
+  };
+
+  const destroyExpression = (destroyPath, callback) => {
+    fetch(destroyPath, {
+      method: 'DELETE',
+      headers: {'X-CSRF-Token': $('meta[name="csrf-token"]').attr('content')}
+    }).then(response => {
+      if (response.ok) {
+        setItemToDelete(null);
+        fetchExpressions(selectedEnvironment, selectedCategory, currentPage);
+      } else {
+        setError(response.statusText);
+      }
+    });
+  };
+
+  useEffect(() => {
+    fetchExpressions(selectedEnvironment, selectedCategory, currentPage);
+  }, [selectedEnvironment, selectedCategory, currentPage]);
+
+  const onEnvironmentSelect = e => {
+    const newSelectedEnvironment = e.target.value;
+    setSelectedEnvironment(newSelectedEnvironment);
+    if (newSelectedEnvironment === DEFAULT_VALUE) {
+      setCategoriesAvailableForSelect(categoriesForSelect);
+    } else {
+      setCategoriesAvailableForSelect(
+        categoriesForSelect.filter(
+          cat => String(cat.envId) === String(newSelectedEnvironment)
+        )
+      );
+      if (String(selectedCategory.envId) !== newSelectedEnvironment) {
+        setSelectedCategory(DEFAULT_VALUE);
+      }
+    }
+    setCurrentPage(1);
```

Comment:
Don't know if this would be better or worse, but I think you can determine when to call `setCurrentPage` using useEffect() with a dependency on `selectedCategory` and `selectedEnvironment`.

Actually, I think you can express the relationship between selectedEnvironment, selectedCategory, and page all through useEffect.  I'm new to this style of writing components though and haven't decided if this is amazing (because it's a lot more declarative) or really hard to follow.

Reply:
To clarify, you're suggesting adding another `useEffect` that is:
```
useEffect(() => {
  setPage(1);
}, [selectedEnvironment, selectedCategory])
```
I think that would work but I'm having trouble justifying why that would be clearer than just doing it in the handler. Plus React handles state updates better in a click handler (related [bug](https://github.com/facebook/react/issues/14259)).

I might be misunderstanding your suggestion though

Reply:
Yes, that was what I was suggesting, but maybe that's not better.  The other thing that comes to mind is that these three pieces of state are actually related so maybe something like `useReducer` would model that more explicitly.  But also ok to leave as-is, up to you.

Reply:
@jamescodeorg I tried using a reducer in df837a06f79 and I'm curious what you think. On one hand, it does move all the logic around the four entangled pieces of state into one place, which is nice. On the other hand, it didn't make sense to me to move all the state to the reducer so now the state is handled in two different ways. I could bring the rest of the state up into the reducer but that state is very simple.

I haven't used `useReducer` before and it doesn't look like it's been used in our codebase so I'm not sure what the conventions here are. It was easy to do (and a good leaning opportunity!) so if it isn't better, I'll just rebase it out.